packages/eslint-plugin/tests/rules/no-unnecessary-coercion.test.ts

```diff
@@ -0,0 +1,212 @@
+import { RuleTester } from '@typescript-eslint/rule-tester';
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+
+import rule from '../../src/rules/no-unnecessary-coercion';
+import { getFixturesRootDir } from '../RuleTester';
+
+const rootDir = getFixturesRootDir();
+
+const ruleTester = new RuleTester({
+  languageOptions: {
+    parserOptions: {
+      tsconfigRootDir: rootDir,
+      project: './tsconfig.json',
+    },
+  },
+});
+
+const messageId = 'unnecessaryCoercion';
+
+ruleTester.run('no-unnecessary-coercion', rule, {
+  valid: [
+    `
+String(1);
+    `,
+    `
+(1).toString();
+    `,
+    `
+\`\${1}\`;
+    `,
+    `
+'' + 1;
+    `,
+    `
+1 + '';
+    `,
+    `
+let str = 1;
+str += '';
+    `,
+    `
+Number('2');
+    `,
+    `
++'2';
+    `,
+    `
+~~'2';
+    `,
+    `
+Boolean(0);
+    `,
+    `
+!!0;
+    `,
+    `
+BigInt(3);
+    `,
+    `
+new String('asdf');
+    `,
+    `
+!false;
+    `,
+    `
+~256;
+    `,
+  ],
+
+  invalid: [
```

Comment:
Couple general nits/info:
1. No need for the `type` in the `error` report. That asserts that the `node` provided in the `context.report()` is the provided type. However, that's useless here, since `node`'s only effect is to provide a report location if `loc` is not provided. And `loc` _is_ provided. So we're just testing anything useful by that 😆 . 
2. Since you're using custom `loc` for the report locations, it's best to specify the full `line`, `endLine`, `column`, endColumn` for the error. We're less careful about that when we just do the `node`'s location but here it makes sense to be more granular.